### PR TITLE
Line strips are no longer a disconnected series of quads

### PR DESCRIPTION
**Pull request by Wumpf**
### What

* Fixes https://github.com/rerun-io/rerun/issues/829
* Fixes https://github.com/rerun-io/rerun/issues/7191

Fixes a long standing issue that the segments of a line have small gaps between them.

Other than originally planned, the solution employed here is entirely based on fragment shader discard + elongation of the line. This can cause some slight artifacts for 3D lines but it's quite hard to spot those issues. Otherwise the biggest drawback of this is that we have more overdraw which gonna be tricky to deal with once we add transparency. On the flip side this is fairly simple and can re-use some of the round cap handling code - round caps no longer use the "trailing triangle" we had for every strip internally, and instead use the quad extension, actually leading to better looking round caps.

⚠️  To allow loops to "just work", all line segments & line strips start and end now with extended round caps. I.e. if your line has a radius of 1 and goes from (0,0,0) to (100,0,0), then what's drawn goes from (-1,0,0) to (101,0,0). Before we didn't have caps on this, so we essentially drew a square starting at (0,0,0) whereas now it's more capsule like (it does not exactly behave like a capsule under perspective though).

Line loops on map view:
Before:
image

After:
image

RRT Star:
Before:
image

After:
image

3D line snippet snippet:
Before:
image

After:
image

Failure Example in #7191
Before - confusingly broken:
image

After - pretty bad rendering performance due to quite massive overdraw of alpha tested geometry, but no longer broken.
image

(After: setting the radius to something large, but sane)
image

re_renderer 2D demo. Left before, right after. Note that "shaded lines" look a lot better now as a sideffect, with shading of arrows slightly broken. Afaik we don't use this anywhere in Rerun today though. 
However, rounded caps are higher quality now in general.

Screenshot 2024-11-11 at 11 10 29

----

Things that were several segments before can now be strips. No visual change there:

Rectangles:
image

Frustum:
image

Box wireframes:

This fixes a regression! In 0.19 those boxes had disconnected lines, in 0.18 this still worked since they were separate segments with rounded corners.

Before (regression):
image

After:
image

----

Due to aforementioned always-round-caps, a list of segments can now look indistinguishable from a single strip:

image

----

Under more extreme circumstances it is still (!) possible to have boxes exhibit some artifacts:

image

This is due to the quasi-capsule nature of our line segments. Calculating a world position from ray-capsule intersection would yield the correct result. What's actually happening right now is that we interpolate the world position over the spanned quad (which is a lot cheaper).

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested the web demo (if applicable):
  * Using examples from latest `main` build: [rerun.io/viewer](https://rerun.io/viewer/pr/8065?manifest_url=https://app.rerun.io/version/main/examples_manifest.json)
  * Using full set of examples from `nightly` build: [rerun.io/viewer](https://rerun.io/viewer/pr/8065?manifest_url=https://app.rerun.io/version/nightly/examples_manifest.json)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG
* [x] If applicable, add a new check to the [release checklist](https://github.com/rerun-io/rerun/blob/main/tests/python/release_checklist)!
* [x] If have noted any breaking changes to the log API in `CHANGELOG.md` and the migration guide

- [PR Build Summary](https://build.rerun.io/pr/8065)
- [Recent benchmark results](https://build.rerun.io/graphs/crates.html)
- [Wasm size tracking](https://build.rerun.io/graphs/sizes.html)

To run all checks from `main`, comment on the PR with `@rerun-bot full-check`.

### Testing
* [x] Tested on WebGL
* [x] Tested on WebGPU

### CODE REVIEW

File: crates/viewer/re_renderer/shader/lines.wgsl [hunk 1]
```diff
@@ -47,6 +47,9 @@ const FLAG_CAP_START_EXTEND_OUTWARDS: u32 = 32u;
 const FLAG_COLOR_GRADIENT: u32 = 64u;
```
Comment (Wumpf) [inline crates/viewer/re_renderer/shader/lines.wgsl:47]: all the really important changes are in this file

File: crates/viewer/re_renderer/shader/composite.wgsl [hunk 2]
```diff
@@ -27,11 +28,17 @@ fn main(in: FragmentInput) -> @location(0) vec4f {
...
     // This is a brittle workaround for the alpha-to-coverage issue described in `ViewBuilder::MAIN_TARGET_ALPHA_TO_COVERAGE_COLOR_STATE`:
     // We need this because otherwise the feathered edges of alpha-to-coverage would be overly bright, as after
     // MSAA-resolve they end up with an unusually low alpha value relative to the color value.
-    color = vec4f(color.rgb * color.a, color.a);
+    if uniforms.blend_with_background == 0 {
+        // To not apply this hack needlessly and account for alpha from alpha to coverage, we have to ignore alpha values if blending is disabled.
+        color = vec4f(color.rgb, 1.0);
+    } else {
+        color = vec4f(color.rgb * color.a, color.a);
+    }
```
Comment (Wumpf) [inline crates/viewer/re_renderer/shader/composite.wgsl:41]: improves artifact suppression. In other words this gets non-map views to where they were in 0.19
Tbh. I'm not 100% sure why I did get still artifacts for boxes in 3D without this, but it also didn't feel _that_ surprising.

**Comment by Wumpf** [comment]
@emilk I should have formulated that more clearly: the _shading_ specifically is slightly broken (see pic). But we don't have the shading ever enabled in Rerun, we made that an opt-in operation unlike on points

**Review by abey79 — APPROVED** [review]
Very nice! Can't say I follow all details in the shader code, but I tested it a bit it seems to work great.

**Comment by emilk** [comment]
Make sure to test this on WebGL _and_ WebGPU!

**Comment by Wumpf** [comment]
> Make sure to test this on WebGL and WebGPU!

Did! On the Mac only mostly though

**Pull request merged**